components/places/src/storage/bookmarks.rs

```diff
@@ -376,6 +395,259 @@ fn delete_bookmark_in_tx(db: &impl ConnExt, guid: &SyncGuid) -> Result<bool> {
+    let existing =
+        get_raw_bookmark(db, guid)?.ok_or_else(|| InvalidPlaceInfo::NoItem(guid.to_string()))?;
+    if existing.bookmark_type != item.bookmark_type() {
+        return Err(InvalidPlaceInfo::InvalidBookmarkType.into());
+    }
+
+    let update_old_parent_status;
+    let update_new_parent_status;
+    // to make our life easier we update every field, using existing when
+    // no value is specified.
+    let parent_id;
+    let position;
+    match item.location() {
+        UpdateTreeLocation::None => {
+            parent_id = existing.parent_id;
+            position = existing.position;
+            update_old_parent_status = false;
+            update_new_parent_status = false;
+        }
+        UpdateTreeLocation::Position(pos) => {
+            parent_id = existing.parent_id;
+            update_old_parent_status = true;
+            update_new_parent_status = false;
+            // Not clear that InvalidParent is the correct error here - probably
+            // should be a "stuff is corrupt" error? Or maybe we should fix it
+            // here by reparenting to unfiled?
+            let parent = get_raw_bookmark(db, &existing.parent_guid)?
+                .ok_or_else(|| InvalidPlaceInfo::InvalidParent(existing.parent_guid.to_string()))?;
+            update_pos_for_deletion(db, existing.position, parent.row_id)?;
+            // We just removed a child, so actual child count is now parent.child_count - 1
+            position = resolve_pos_for_insert(db, pos, parent.row_id, parent.child_count - 1)?;
+        }
+        UpdateTreeLocation::Parent(new_parent_guid, pos) => {
+            if BookmarkRootGuid::from_guid(&new_parent_guid) == Some(BookmarkRootGuid::Root) {
+                return Err(InvalidPlaceInfo::InvalidGuid.into());
+            }
+            let new_parent = get_raw_bookmark(db, &new_parent_guid)?
+                .ok_or_else(|| InvalidPlaceInfo::InvalidParent(new_parent_guid.to_string()))?;
+            if new_parent.bookmark_type != BookmarkType::Folder {
+                return Err(InvalidPlaceInfo::InvalidParent(new_parent_guid.to_string()).into());
+            }
+            parent_id = new_parent.row_id;
+            update_old_parent_status = true;
+            update_new_parent_status = true;
+            // This position change is more complicated across parents.
+            // As above, this failure really means "stuff is corrupt" and we
+            // could consider fixing the tree instead of throwing?
+            let existing_parent = get_raw_bookmark(db, &existing.parent_guid)?
+                .ok_or_else(|| InvalidPlaceInfo::InvalidParent(existing.parent_guid.to_string()))?;
+            update_pos_for_deletion(db, existing.position, existing_parent.row_id)?;
+            position = resolve_pos_for_insert(db, pos, new_parent.row_id, new_parent.child_count)?;
+        }
+    };
+    let place_id = match item {
+        UpdatableItem::Bookmark(b) => match &b.url {
```

Comment:
I think we need to check and error if the URL is too long here (The `*_page_info` functions might do it though, I don't remember).

Reply:
thanks - new_page_info() makes that check (not that I had considered that before your comment!)